src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java

```diff
@@ -31,68 +35,87 @@
 import java.util.Map;
 import java.util.TreeMap;
 
-import org.rumbledb.api.Item;
-
-public class VariableReferenceIterator extends LocalRuntimeIterator {
+public class VariableReferenceIterator extends HybridRuntimeIterator {
 
 
     private static final long serialVersionUID = 1L;
-    private SequenceType sequence;
+    private SequenceType _sequence;
     private String _variableName;
-    private List<Item> items = null;
-    private int currentIndex = 0;
+    private List<Item> _items = null;
+    private int _currentIndex = 0;
 
     public VariableReferenceIterator(String variableName, SequenceType seq, IteratorMetadata iteratorMetadata) {
         super(null, iteratorMetadata);
-        this._variableName = "$" + variableName;
-        this.sequence = seq;
+        _variableName = "$" + variableName;
+        _sequence = seq;
+    }
+
+    @Override
+    protected boolean initIsRDD() {
+        // TODO: figure out how to set isRDD for varRefIter.
+        //return _currentDynamicContext.isRDD(_variableName, getMetadata());
+        return false;
     }
 
     @Override
-    public Item next() {
-        if (!this.hasNext())
+    public JavaRDD<Item> getRDDAux(DynamicContext context) {
+        return _currentDynamicContext.getRDDVariableValue(_variableName, getMetadata());
+    }
+
+    @Override
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        return _currentDynamicContext.getDFVariableValue(_variableName, getMetadata());
+    }
+
+    @Override
+    protected boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    public Item nextLocal() {
+        if (!_hasNext) {
             throw new IteratorFlowException(
                     RuntimeIterator.FLOW_EXCEPTION_MESSAGE + "$" + _variableName,
                     getMetadata()
             );
-        Item item = items.get(currentIndex);
-        currentIndex++;
-        if (currentIndex == items.size())
-            this._hasNext = false;
+        }
+        Item item = _items.get(_currentIndex);
+        _currentIndex++;
+        if (_currentIndex == _items.size()) {
+            _hasNext = false;
+        }
         return item;
     }
 
     @Override
-    public void reset(DynamicContext context) {
-        super.reset(context);
-        this.currentIndex = 0;
-        this.items = null;
+    public void openLocal() {
+        _currentIndex = 0;
+        _items = _currentDynamicContext.getLocalVariableValue(_variableName, getMetadata());
```

Comment:
Thank you. Could you help me understand why, here, it does not automatically materialize the RDD inside the dynamic context?

I remember that we had implemented these methods as an automatic cascade, i.e., the dynamic context would automatically materialize the value if it is an RDD. Am I missing something?

Reply:
Could you take a look at the CountFunctionIterator? I believe that may answer your question as the execution in the test case flows through there. 

Ps: I have to run somewhere now, we can pick up this discussion on Monday when you're available. Have a nice weekend!

Reply:
Happy to do so. Have a nice weekend!

Reply:
(And for taking it from there on Monday, here is my fix: https://github.com/RumbleDB/rumble/pull/384 )